ext/standard/math.c

```diff
@@ -95,22 +95,75 @@ static inline double php_intpow10(int power) {
 static inline double php_round_helper(double value, int mode) {
 	double tmp_value;
 
-	if (value >= 0.0) {
-		tmp_value = floor(value + 0.5);
-		if ((mode == PHP_ROUND_HALF_DOWN && value == (-0.5 + tmp_value)) ||
-			(mode == PHP_ROUND_HALF_EVEN && value == (0.5 + 2 * floor(tmp_value/2.0))) ||
-			(mode == PHP_ROUND_HALF_ODD  && value == (0.5 + 2 * floor(tmp_value/2.0) - 1.0)))
-		{
-			tmp_value = tmp_value - 1.0;
-		}
-	} else {
-		tmp_value = ceil(value - 0.5);
-		if ((mode == PHP_ROUND_HALF_DOWN && value == (0.5 + tmp_value)) ||
-			(mode == PHP_ROUND_HALF_EVEN && value == (-0.5 + 2 * ceil(tmp_value/2.0))) ||
-			(mode == PHP_ROUND_HALF_ODD  && value == (-0.5 + 2 * ceil(tmp_value/2.0) + 1.0)))
-		{
-			tmp_value = tmp_value + 1.0;
-		}
+	switch (mode) {
+		case PHP_ROUND_HALF_UP:
+			if (value >= 0.0) {
+				tmp_value = floor(value + 0.5);
+			} else {
+				tmp_value = ceil(value - 0.5);
+			}
+			break;
+		case PHP_ROUND_HALF_DOWN:
+			if (value >= 0.0 ) {
+				tmp_value = floor(value + 0.5);
+				if(value + 0.5 == tmp_value) {
+					tmp_value = tmp_value - 1;
+				}
+			} else {
+				tmp_value = ceil(value - 0.5);
+				if(value - 0.5 == tmp_value) {
+					tmp_value = tmp_value + 1.0;
+				}
```

Comment:
Likewise `tmp_value = floor(value + 0.5)`.

Reply:
Similarly here. Values in the range -0.5 and 0 will be rounded to positive zero, not negative. This will be inconsistent with the behaviour of function `floor()`. Personally I would change all negative zeros to positive, but this would require also changes to `ceil()` and `floor()` functions.